core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java

```diff
@@ -0,0 +1,287 @@
+        }
+    }
+
+    public void testConcurrentCancel() throws Exception {
+        final int numThreads = scaledRandomIntBetween(2, 8);
+        final CountDownLatch latch = new CountDownLatch(numThreads + 1);
+        final AtomicInteger counter = new AtomicInteger(0);
+        Runnable runnable = counter::incrementAndGet;
+
+        final Cancellable cancellable = threadPool.scheduleWithFixedDelay(runnable, TimeValue.timeValueSeconds(30L), Names.GENERIC);
+        List<Thread> threadList = new ArrayList<>(numThreads);
+        for (int i = 0; i < numThreads; i++) {
+            threadList.add(new Thread() {
+                @Override
+                public void run() {
+                    latch.countDown();
+                    try {
+                        latch.await();
+                        cancellable.cancel();
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    }
+                }
+            });
+        }
+
+        for (Thread thread : threadList) {
+            thread.start();
+        }
+        latch.countDown();
+        for (Thread thread : threadList) {
+            thread.join();
+        }
+
+        assertTrue(cancellable.isCancelled());
+    }
+
+    public void testBlockingCallOnSchedulerThreadFails() throws Exception {
+        final AtomicReference<Throwable> throwableRef = new AtomicReference<>();
+        final BaseFuture<Object> future = new BaseFuture<Object>() {};
+        final Runnable runnable = () -> {
+            try {
+                if (randomBoolean()) {
+                    future.get();
+                } else {
+                    future.get(1L, TimeUnit.SECONDS);
+                }
+            } catch (Throwable t) {
+                throwableRef.set(t);
+            }
+        };
+
+        Cancellable cancellable = threadPool.scheduleWithFixedDelay(runnable, TimeValue.timeValueMillis(100L), Names.SAME);
+        boolean failed = awaitBusy(() -> throwableRef.get() != null, 1L, TimeUnit.SECONDS);
+        assertTrue(failed);
+        assertThat(throwableRef.get(), instanceOf(AssertionError.class));
+        assertThat(throwableRef.get().getMessage(), containsString("Blocking"));
+        assertFalse(cancellable.isCancelled());
+    }
+
+    public void testBlockingCallOnNonSchedulerThreadAllowed() throws Exception {
+        class TestFuture extends BaseFuture<Object> {
+            public boolean futureDone(Object value) {
+                return set(value);
+            }
+        }
+
+        final AtomicReference<Throwable> throwableRef = new AtomicReference<>();
+        final AtomicReference<Object> objectRef = new AtomicReference<>();
+        final TestFuture future = new TestFuture();
+        final boolean rethrow = randomBoolean();
+
+        final Runnable runnable = () -> {
+            try {
+                if (randomBoolean()) {
```

Comment:
can we fix this into a boolean before we start? will be easier to reproduce / know what happened.